nixpkgs/toolchains/rust.bzl

```diff
@@ -0,0 +1,159 @@
+                ],
+                # Some patterns (e.g. `lib/*.a`) don't match anything, see https://github.com/bazelbuild/rules_rust/pull/245
+                allow_empty = True,
+            ),
+            visibility = ["//visibility:public"],
+        )
+
+        filegroup(
+            name = "rust_doc",
+            srcs = ["bin/rustdoc"],
+            visibility = ["//visibility:public"],
+        )
+
+        load('@rules_rust//rust:toolchain.bzl', 'rust_toolchain')
+        rust_toolchain(
+            name = "rust_nix_impl",
+            rustc = ":rustc",
+            rustc_lib = ":rustc_lib",
+            rust_lib = ":rust_lib",
+            rust_doc = ":rust_doc",
+            binary_ext = "",
+            staticlib_ext = ".a",
+            dylib_ext = ".so",
+            stdlib_linkflags = ["-lpthread", "-ldl"],
+            os = "${os}",
+            target_triple = "${target}",
+        )
+        EOF
+    '';
+}
+"""
+
+_rust_nix_toolchain = """
+toolchain(
+    name = "rust_nix",
+    toolchain = "@{toolchain_repo}//:rust_nix_impl",
+    toolchain_type = "@rules_rust//rust:toolchain",
+    exec_compatible_with = {exec_constraints},
+    target_compatible_with = {target_constraints},
+)
+"""
+
+def _ensure_constraints(repository_ctx):
+    cpu = get_cpu_value(repository_ctx)
+    os = {"darwin": "osx"}.get(cpu, "linux")
+    if not repository_ctx.attr.target_constraints and not repository_ctx.attr.exec_constraints:
+        target_constraints = ["@platforms//cpu:x86_64"]
+        target_constraints.append("@platforms//os:{}".format(os))
+        exec_constraints = target_constraints
+    else:
+        target_constraints = list(repository_ctx.attr.target_constraints)
+        exec_constraints = list(repository_ctx.attr.exec_constraints)
+    exec_constraints.append("@io_tweag_rules_nixpkgs//nixpkgs/constraints:support_nix")
+    return exec_constraints, target_constraints
+
+def _nixpkgs_rust_toolchain_impl(repository_ctx):
+    exec_constraints = [ "@io_tweag_rules_nixpkgs//nixpkgs/constraints:support_nix", "@platforms//cpu:x86_64" ]
+    target_constraints = [ "@platforms//cpu:x86_64" ]
+    cpu = get_cpu_value(repository_ctx)
+    repository_ctx.file(
+            "BUILD.bazel",
+            executable = False,
+            content = _rust_nix_toolchain.format(
+                    toolchain_repo = repository_ctx.attr.toolchain_repo,
+                    exec_constraints = exec_constraints,
+                    target_constraints = target_constraints
+            )
+    )
+
+_nixpkgs_rust_toolchain = repository_rule(
+    _nixpkgs_rust_toolchain_impl,
+    attrs = {
+        "toolchain_repo": attr.string(),
+    },
+)
+
+def nixpkgs_rust_configure(
+        sdk_name = "rust_linux_x86_64",
+        repository = None,
+        repositories = {},
+        nix_file = None,
+        nix_file_deps = None,
+        nix_file_content = None,
+        nixopts = [],
+        fail_not_supported = True,
+        quiet = False,
```

Comment:
Regarding exec and target platform constraints. You could also follow the pattern used e.g. [here](https://github.com/tweag/rules_nixpkgs/blob/05a445575de51872f528c537cc7ff3ab11114f21/nixpkgs/nixpkgs.bzl#L1114-L1115) and then [here](https://github.com/tweag/rules_nixpkgs/blob/05a445575de51872f528c537cc7ff3ab11114f21/nixpkgs/nixpkgs.bzl#L1028) to let the user configure these.

In the simple case [`_ensure_constraint`](https://github.com/tweag/rules_nixpkgs/blob/05a445575de51872f528c537cc7ff3ab11114f21/nixpkgs/nixpkgs.bzl#L569) would pick the current os as the default. That said, aarch64 support is not currently implemented in that approach.

In that case it would make sense to choose a more generic default name like `nixpkgs_rust_config`.

Happy to leave this for a follow-up PR.

Reply:
I've added user-configurable parameters to both Rust and foreign-cc toolchains. I'm not sure what you mean by a more generic default name.